File: MPT_Monte_Carlo.py
"""This program utilizes a Monte Carlo Simulation to create a means of comparison
for investing $100,000 in (1) an aggressive portfolio and (2) a conservative portfolio"""
import numpy as np

mean_ag = 9.4324/100
std_ag = 15.675/100
mean_con = 6.189/100
std_con = 6.3438/100
years = 20 # 20 year simulation
inflation = 0.035 # annual inflation is 3.5%
trials = 10000
initial = 100000

def monte_carlo_simulation(intial, trials, mean, std):
    """Runs a Monte Carlo Simulation of a portfolio with MEAN and STD for TRIALS number of times.
    Returns inflation-adjusted list of returns (performances) for each simluation"""
    performances = list()
    for _ in range(trials):
        performances.append(run_simulation(initial, years, mean, std))
    return performances

def run_simulation(initial, years, mean, std):
    """Returns the inflation-adjusted returns of a portfolio with MEAN and STD
    over a YEARS period"""
    for _ in range(years):
        initial *= 1 + np.random.normal(mean, std)
    return inflation_adjuster(initial, years, inflation)

def inflation_adjuster(future_value, years, inflation):
    """Accounts for inflation to adjust future dollars into present value dollars"""
    return future_value / (1 + inflation)**years

# Keeping track of the returns of each 20 year simlulation
performances_ag = monte_carlo_simulation(initial, trials, mean_ag, std_ag)
performances_con = monte_carlo_simulation(initial, trials, mean_con, std_con)

std_performance_ag = np.std(performances_ag)
std_performance_con = np.std(performances_con)

#Median 20 years performance
median_performance_ag = np.median(performances_ag)
median_performance_con = np.median(performances_con)

#10% best case (90th percentile)
ninetieth_percentile_ag = np.percentile(performances_ag, 90)
ninetieth_percentile_con = np.percentile(performances_con, 90)

#10% worse case (10th percentile)
tenth_percentile_ag = np.percentile(performances_ag, 10)
tenth_percentile_con = np.percentile(performances_con, 10)

def extract_info(portfolio, median, ninetieth, tenth):
    return f"Portfolio type: {portfolio} \nMedian 20 Yr: {median} \n" \
            f"10% best case: {ninetieth} \n10% worse case: {tenth}"
print(extract_info("Aggresive", median_performance_ag, ninetieth_percentile_ag,
                    tenth_percentile_ag))
print("\n")
print(extract_info("Very Conservative", median_performance_con,
                    ninetieth_percentile_con, tenth_percentile_con))

File: MPTMonteCarlo.java
/* This program utilizes a Monte Carlo Simulation to create a means of comparison
for investing $100,000 in (1) an aggressive portfolio and (2) a conservative portfolio */

import java.util.Arrays;
import java.util.Random;
import java.util.HashMap;

public class MPTMonteCarlo{
  static double meanAggressive = 9.4324/100;
  static double stdAggressive = 15.675/100;
  static double meanConservative = 6.189/100;
  static double stdConservative = 6.3438/100;
  static int years = 20; // 20 year simulation
  static double inflation = 0.035; // annual inflation is 3.5%
  static int trials = 10000;
  static double initial = 100000;

  public static void main(String[] args){
    HashMap<String, Double> informationAggressive = extractInfo(initial, trials, meanAggressive, stdAggressive);
    HashMap<String, Double> informationConservative = extractInfo(initial, trials, meanConservative, stdConservative);
    System.out.println(readInfo("Aggressive", informationAggressive));
    System.out.println();
    System.out.println(readInfo("Very Conservative", informationConservative));
  }

  static HashMap<String, Double> extractInfo(double initial, int trials, double mean, double std){
    /** Given the INTIAL value of the portfolio, the numbers of TRIALS, the MEAN return of the portfolio
    and the STD (standard deviation), extracts the Median, NinetyPercentile, and TenthPercentile and Returns
    relevant information in a HashMap */
    HashMap<String, Double> information = new HashMap<String, Double>();
    double[] sortedPerformances = monteCarloSimluations(initial, trials, mean, std);
    information.put("STD", getSTD(sortedPerformances));
    information.put("Median", getPercentile(sortedPerformances, 50));
    information.put("NinetyPercentile", getPercentile(sortedPerformances, 90));
    information.put("TenthPercentile", getPercentile(sortedPerformances, 10));
    return information;
  }

  static String readInfo(String portfolio, HashMap<String, Double> information){
    /** Given the PORTOFLIO type and relevant INFORMATION about the portoflio, returns a string
    displaying the relevant finanical metrics */
    return String.format("Portfolio type: %s \nMedian 20 Yr: %f \n10%% best case: %f \n10%% worse case: %f",
    portfolio, information.get("Median"), information.get("NinetyPercentile"), information.get("TenthPercentile"));
  }

  static double[] monteCarloSimluations(double initial, int trials, double mean, double std){
    /** Runs a Monte Carlo Simulation of a portfolio with INITIAL value and with MEAN and STD for TRIALS
    number of times. Returns inflation-adjusted array of returns (performances) for each simluation */
    double[] performances = new double[trials];
    for (int i = 0; i < trials; i++){
      performances[i] = (runSimulation(initial, years, mean, std));
    }
    Arrays.sort(performances);
    return performances;
  }

  static double runSimulation(double initial, int years, double mean, double std){
    /** Returns the inflation-adjusted returns of a portfolio with MEAN and STD
    over a YEARS period */
    for (int i = 0; i < years; i++){
      initial *= 1 + drawGaussianSample(mean, std);
    }
    return inflationAdjuster(initial, years, inflation);
  }

  static double drawGaussianSample(double mean, double std){
    /** Selects a return for next year by drawing randomly from a Gaussian Distribution
    of potenital returns given MEAN and STD of the portfolio */
    Random r = new Random();
    return (r.nextGaussian() * std) + mean;
  }

  static double inflationAdjuster(double futureValue, int years, double inflation){
    /** Adjusts futureValue of money to presentValue of money */
    return futureValue / Math.pow((1 + inflation), years);
  }

  static double getMean(double[] performances){
    /** Given a list PERFORMANCES of end portfolio values, returns the mean */
    double sum = 0;
    for (double performance : performances){
      sum += performance;
    }
    return sum / performances.length;
  }

  static double getSTD(double[] performances){
    /** Given a list PERFORMANCES of end portfolio values, returns the standard deviation */
    double variance = 0;
    double mean = getMean(performances);
    for (double performance : performances){
      variance += Math.pow((mean - performance), 2);
    }
    variance = variance / performances.length;
    return Math.sqrt(variance);
  }

  static double getPercentile(double[] performances, double percentile){
    /** Given a list PERFORMANCES of end portfolio values and the desired PERCENTILE,
     returns the value corresponding to that PERCENTILE */
    int index;
    double closest = (percentile / 100.0) * performances.length;
    index = (int) Math.ceil(closest);
    return performances[index];
  }

}
